crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4567,6 +4567,13 @@ impl<'db> TypeInferenceBuilder<'db> {
                 let param_type = self.infer_type_expression(parameters);
                 UnionType::from_elements(self.db, [param_type, Type::none(self.db)])
             }
+            KnownInstanceType::Union => match parameters {
+                ast::Expr::Tuple(t) => UnionType::from_elements(
+                    self.db,
+                    t.iter().map(|elt| self.infer_type_expression(elt)),
+                ),
+                _ => self.infer_type_expression(parameters),
+            },
```

Comment:
This is extremely subtle, but with this change, we now fail to infer a type for the `ast::Expr::Tuple` expression itself. We only infer types for the sub-expressions via `self.infer_type_expression(elt)`, but we don't store a type for the overall `ast::Expr::Tuple` (only for the whole slice expression).

In other words, when we have something like `Union[str, int]`, we do infer types for the whole slice expression `Union[str, int]`. We also infer types for `int` and `str`, but we do not infer/store a type for the `int, str` tuple expression.

Unfortunately, the test that would have caught this is currently deactivated (see #14391), but I hope we can re-activate it soon.

I'll open a PR with a fix.